src/config.ts

```diff
@@ -14,19 +14,23 @@ export function update<T>(key: string, value: T,
 }
 
 // Traverse a JSON value, replacing placeholders in all strings.
-function substitute<T>(val: T): T {
+async function substitute<T>(val: T): Promise<T> {
   if (typeof val === 'string') {
-    val = val.replace(/\$\{(.*?)\}/g, (match, name) => {
-      // If there's no replacement available, keep the placeholder.
-      return replacement(name) ?? match;
-    }) as unknown as T;
+    const replacementPattern = /\$\{(.*?)\}/g;
+    const replacementPromises: Promise<string | undefined>[] = [];
+    val.replace(replacementPattern, (match, name) => {
+      replacementPromises.push(replacement(name));
+      return match;
+    });
+    const replacements = await Promise.all(replacementPromises);
+    val = val.replace(/\$\{(.*?)\}/g, (match, _) => replacements.shift() ?? match) as unknown as T;
```

Comment:
* Use `replacementPattern` rather than repeating the regex
* Preserve the comment `// If there's no replacement available, keep the placeholder.`, which explains why we do the `?? match` part.
* nit: the function can be written a bit more simply as`match => replacements.shift() ?? match`

So, putting these together, suggested rewrite:

```ts
val = val.replace(replacementPattern,
    // If there's no replacement available, keep the placeholder.
    match => replacements.shift() ?? match) as unknown as T;
```

Reply:
Updated